nomad/host_volume_endpoint.go

```diff
@@ -0,0 +1,515 @@
+	// atomic with the registration, and creating the volume provides values we
+	// want to write on the Volume in raft anyways.
+
+	// This can't reuse the validVols slice because we only want to write
+	// volumes we've successfully created or updated on the client to get
+	// updated in Raft.
+	raftArgs := &structs.HostVolumeRegisterRequest{
+		Volumes:      []*structs.HostVolume{},
+		WriteRequest: args.WriteRequest,
+	}
+
+	var mErr *multierror.Error
+	for _, vol := range validVols {
+		err = v.createVolume(vol) // mutates the vol
+		if err != nil {
+			mErr = multierror.Append(mErr, err)
+		} else {
+			raftArgs.Volumes = append(raftArgs.Volumes, vol)
+		}
+	}
+
+	// if we created or updated any volumes, apply them to raft.
+	var index uint64
+	if len(raftArgs.Volumes) > 0 {
+		_, index, err = v.srv.raftApply(structs.HostVolumeRegisterRequestType, raftArgs)
+		if err != nil {
+			v.logger.Error("raft apply failed", "error", err, "method", "register")
+			mErr = multierror.Append(mErr, err)
+		}
+	}
+
+	reply.Volumes = raftArgs.Volumes
+	reply.Index = index
+	return helper.FlattenMultierror(mErr)
+}
+
+func (v *HostVolume) Register(args *structs.HostVolumeRegisterRequest, reply *structs.HostVolumeRegisterResponse) error {
+
+	authErr := v.srv.Authenticate(v.ctx, args)
+	if done, err := v.srv.forward("HostVolume.Register", args, args, reply); done {
+		return err
+	}
+	v.srv.MeasureRPCRate("host_volume", structs.RateMetricWrite, args)
+	if authErr != nil {
+		return structs.ErrPermissionDenied
+	}
+	defer metrics.MeasureSince([]string{"nomad", "host_volume", "register"}, time.Now())
+
+	allowVolume := acl.NamespaceValidator(acl.NamespaceCapabilityHostVolumeRegister)
+	aclObj, err := v.srv.ResolveACL(args)
+	if err != nil {
+		return err
+	}
+
+	if len(args.Volumes) == 0 {
+		return fmt.Errorf("missing volume definition")
+	}
+
+	for _, vol := range args.Volumes {
+		if vol.Namespace == "" {
+			vol.Namespace = args.RequestNamespace()
+		}
+		if !allowVolume(aclObj, vol.Namespace) {
+			return structs.ErrPermissionDenied
+		}
+	}
+
+	// ensure we only try to create valid volumes or make valid updates to
+	// volumes
+	validVols, err := v.validateVolumeUpdates(args.Volumes)
+	if err != nil {
+		return err
+	}
+
+	raftArgs := &structs.HostVolumeRegisterRequest{
+		Volumes:      validVols,
+		WriteRequest: args.WriteRequest,
+	}
+
+	var mErr *multierror.Error
+	var index uint64
+	if len(raftArgs.Volumes) > 0 {
+		_, index, err = v.srv.raftApply(structs.HostVolumeRegisterRequestType, raftArgs)
+		if err != nil {
+			v.logger.Error("raft apply failed", "error", err, "method", "register")
+			mErr = multierror.Append(mErr, err)
+		}
+	}
+
+	reply.Volumes = raftArgs.Volumes
+	reply.Index = index
+	return helper.FlattenMultierror(mErr)
+}
+
+func (v *HostVolume) validateVolumeUpdates(requested []*structs.HostVolume) ([]*structs.HostVolume, error) {
+
+	now := time.Now().UnixNano()
+
+	vols := []*structs.HostVolume{}
```

Comment:
more IDE complaining

```suggestion
	var vols []*structs.HostVolume
```